File: Live-Streaming/Agora-Interactive-Broadcasting-Live-Streaming-Android/app/src/main/java/io/agora/interactivebroadcastingwithcdnstreaming/MessageItemDecoration.java
package io.agora.interactivebroadcastingwithcdnstreaming;

import android.graphics.Rect;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;

public class MessageItemDecoration extends RecyclerView.ItemDecoration {
    private static final int DIVIDER = 16;
    private static final int HEADER = 3;
    private static final int FOOTER = 3;

    @Override
    public void getItemOffsets(@NonNull Rect outRect, @NonNull View view, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        super.getItemOffsets(outRect, view, parent, state);

        int itemCount = parent.getAdapter().getItemCount();
        int viewPosition = parent.getChildAdapterPosition(view);

        outRect.left = DIVIDER;
        outRect.right = DIVIDER;
        if (viewPosition == 0) {
            outRect.top = HEADER;
            outRect.bottom = DIVIDER / 2;
        } else if (viewPosition == itemCount - 1) {
            outRect.top = DIVIDER / 2;
            outRect.bottom = FOOTER;
        } else {
            outRect.top = DIVIDER / 2;
            outRect.bottom = DIVIDER / 2;
        }

    }
}

File: Live-Streaming/Agora-Interactive-Broadcasting-Live-Streaming-Android-Kotlin/app/src/main/java/io/agora/interactivebroadcastingwithcdnstreaming/MessageItemDecoration.kt
package io.agora.interactivebroadcastingwithcdnstreaming

import android.graphics.Rect
import android.support.v7.widget.RecyclerView
import android.view.View

class MessageItemDecoration : RecyclerView.ItemDecoration() {

    companion object {
        private const val DIVIDER = 16
        private const val HEADER = 3
        private const val FOOTER = 3
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
        super.getItemOffsets(outRect, view, parent, state)

        val itemCount = parent.adapter!!.itemCount
        val viewPosition = parent.getChildAdapterPosition(view)

        outRect.left = DIVIDER
        outRect.right = DIVIDER
        when (viewPosition) {
            0 -> {
                outRect.top = HEADER
                outRect.bottom = DIVIDER / 2
            }
            itemCount - 1 -> {
                outRect.top = DIVIDER / 2
                outRect.bottom = FOOTER
            }
            else -> {
                outRect.top = DIVIDER / 2
                outRect.bottom = DIVIDER / 2
            }
        }

    }
}
